### PR TITLE
[TG-1358] Fix support for functions containing more than 255 local variables

### PR DESCRIPTION
**Pull request by thk123**
Wide `iload` commands were incorrectly only reading a single byte for the local variable index. 

Still to do:

 - unit test for this behaviour to make the test more robust and faster (currently takes 7 seconds even though the problem is in the first phase)

### CODE REVIEW

File: regression/cbmc-java/lots_of_local_variables/TooManyLocals.java [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+    allInts[249] = i249;
+    allInts[250] = i250;
+    allInts[251] = i251;
+    allInts[252] = i252;
+    allInts[253] = i253;
+    allInts[254] = i254;
+    allInts[255] = i255;
+    assert(allInts[255] + allInts[0] == 255);
+  }
+}
```
Comment (thk123) [inline regression/cbmc-java/lots_of_local_variables/TooManyLocals.java:518]: Missing new line

File: regression/cbmc-java/lots_of_local_variables/test.desc [hunk 2]
```diff
@@ -0,0 +1,6 @@
+CORE
+TooManyLocals.class
+--function TooManyLocals.test
+VERIFICATION SUCCESSFUL
+\[java::TooManyLocals\.test:\(\)V\.assertion\.1\] .*: SUCCESS
+EXIT=0
```
Comment (thk123) [inline regression/cbmc-java/lots_of_local_variables/test.desc:6]: Missing new line

**Review by mgudemann — APPROVED** [review]
I think a proper crafted class file without arrays would be much simpler

**Review by smowton — COMMENTED** [review]
Is wide instruction mutually exclusive with the other format characters, or are those invariants acknowledging a shortcoming?

**Comment by thk123** [comment]
@smowton I'm not sure - they are guards to find the cases where it would be a shortcoming, or else documenting that these formats cannot be wide. I'd rather not at this point investigate for all of them whether they can be wide or not. But with these guards if we do find code that has them we'll now get a clear message explaining what has gone wrong.

**Comment by thk123** [comment]
@smowton I checked and the only formats that can be wide modified are 'v' and 'V', would you like me to change the invariant in some way?

File: regression/cbmc-java/lots_local_variables_manual/test.desc [hunk 3]
```diff
@@ -0,0 +1,4 @@
+CORE
+ManyLocalVariables.class
+~/workspace/test-gen/build/release/bin/jbmc ManyLocalVariables.class --lazy-methods --function ManyLocalVariables.test
+VERIFCATION SUCCESSFUL
```
Comment (mgudemann) [inline regression/cbmc-java/lots_local_variables_manual/test.desc]: misssing newline

**Comment by smowton** [comment]
Would it be simplest then to say `if(wide) invariant(format == 'v' || format == 'V')`?

**Review by smowton — APPROVED** [review]
Approving anyhow, happy for this to go in either way

File: regression/cbmc-java/lots_of_local_variables/TooManyLocals.java [hunk 4]
```diff
@@ -0,0 +1,520 @@
+// import java.util.Random;
```
Comment (thk123) [inline regression/cbmc-java/lots_of_local_variables/TooManyLocals.java]: Remove unused import

File: regression/cbmc-java/lots_local_variables_manual/test.desc [hunk 5]
```diff
@@ -0,0 +1,4 @@
+CORE
+ManyLocalVariables.class
+--lazy-methods --function ManyLocalVariables.test
```
Comment (thk123) [inline regression/cbmc-java/lots_local_variables_manual/test.desc]: Presumably lazy methods isn't needed here?

File: src/java_bytecode/java_bytecode_parser.cpp [hunk 6]
```diff
@@ -773,10 +780,20 @@ void java_bytecode_parsert::rbytecode(
...
+        }
+        else
+        {
+          u1 v = read_u1();
+          instruction.args.push_back(from_integer(v, unsignedbv_typet(8)));
+          address += 1;
+        }
       }
-      address+=1;
+
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_parser.cpp:798]: Remove blank line

**Review by thk123 — COMMENTED** [review]
Is there value in having a "real" Java example ontop of the Jasmin generated class file?

File: regression/cbmc-java/lots_local_variables_manual/test.desc [hunk 7]
```diff
@@ -0,0 +1,4 @@
...
+VERIFCATION SUCCESSFUL
```
Comment (peterschrammel) [inline regression/cbmc-java/lots_local_variables_manual/test.desc]: typo (causes https://travis-ci.org/diffblue/cbmc/jobs/339103409#L2618): VERIFICATION

**Comment by mgudemann** [comment]
@thk123 should be mergeable now

**Comment by thk123** [comment]
Squashed @mgudemann fixes (thanks!) will merge once passes CI

**Pull request merged**